### ISSUE TITLE
Stackoverfow on deeply nested join/map

### ISSUE THREAD

**Issue by weiznich** [alexcrichton/futures-rs#202]
The following code will trigger a stack overflow.

``` rust
extern crate futures;

use futures::{Future, done};

const FUTURE_COUNT: usize = 10_000;

fn main() {
    let f = (0..FUTURE_COUNT)
        .map(|_| done::<i32, i32>(Result::Ok(42)))
        .fold(done(Result::Ok(0)).boxed(),
              |acc, f| acc.join(f).map(|(a, b)| a + b).boxed());

    let r = ::futures::task::spawn(f).wait_future();
    assert_eq!(r, Ok(42 * FUTURE_COUNT as i32));
    println!("Hello, world!");
}

```

[gdb backtrace](https://gist.github.com/weiznich/1e17dcce6eacdb5c145be11e404ee61f)

**Comment by alexcrichton** [comment]
Thanks for the report! This is indeed the case where massively recursive futures like this may stack overflow. This is similar to massively recursive functions in Rust, however, which may also stack overflow. Unfortunately this is likely to be a "WONTFIX" issue in the sense that the design of the `Future` trait precludes any mitigation being done here.

Note that while developing this library this was an issue that we were quite aware of. Initially we tried mitigations through trait methods and such but unfortunately it never panned out, it ended up always imposing a cost on downstream consumers.

Another important point we found, though, is that almost all examples which lead to this sort of recursion can typically be written in such a way to avoid it. For example here instead of 10K `Join` futures it'd be far more efficient to, for example, use the `collect` future (and this would also avoid the stack overflow problem).

Does that make sense? Did you hit this particular case n the wild though where it was tough to work around?
